src/compiler/checker.ts

```diff
@@ -4929,6 +4998,12 @@ module ts {
                 if (source === anyFunctionType) {
                     return;
                 }
+                if (target.flags & TypeFlags.TypeGuard) {
+                    target = (<TypeGuardType>target).type;
+                }
+                if (source.flags & TypeFlags.TypeGuard) {
+                    source = (<TypeGuardType>source).type;
+                }

```

Comment:
It infers a boolean because `x is Foo` is of type boolean. It is not a question of assignability to T, you are actually inferring what type T should be. And in this case it is the type given by `x is Foo` which is boolean.

For your examples, you are correct, they should not work as type guards. The reason is that in both of them `x is Foo` is actually a statement about `p` in the context of the call to `generic`. And there is no way to know what `generic` passes for `p` internally (as an argument to `func`). In particular, we have no idea if the argument bears any relation to `a`, which is the thing you're trying to narrow.